stuffer/s2n_stuffer_base64.c

```diff
@@ -13,124 +13,71 @@
-             */
-            POSIX_ENSURE(!(o.data[3] != '=' || value2 & 0x0f), S2N_ERR_INVALID_BASE64);
-            bytes_this_round = 1;
-            value3 = 0;
-            value4 = 0;
-        } else if (o.data[3] == '=') {
-            /* The last two bits of the final value should be unset */
-            POSIX_ENSURE(!(value3 & 0x03), S2N_ERR_INVALID_BASE64);
-
-            bytes_this_round = 2;
-            value4 = 0;
-        }
-
-        /* Advance by bytes_this_round, and then fill in the data */
-        POSIX_GUARD(s2n_stuffer_skip_write(out, bytes_this_round));
-        uint8_t *ptr = out->blob.data + out->write_cursor - bytes_this_round;
-
-        /* value1 maps to the first 6 bits of the first data byte */
-        /* value2's top two bits are the rest */
-        *ptr = ((value1 << 2) & 0xfc) | ((value2 >> 4) & 0x03);
-
-        if (bytes_this_round > 1) {
-            /* Put the next four bits in the second data byte */
-            /* Put the next four bits in the third data byte */
-            ptr++;
-            *ptr = ((value2 << 4) & 0xf0) | ((value3 >> 2) & 0x0f);
-        }
-
-        if (bytes_this_round > 2) {
-            /* Put the next two bits in the third data byte */
-            /* Put the next six bits in the fourth data byte */
-            ptr++;
-            *ptr = ((value3 << 6) & 0xc0) | (value4 & 0x3f);
-        }
-    } while (bytes_this_round == 3);
+
+    int base64_groups = s2n_stuffer_data_available(stuffer) / 4;
+    if (base64_groups == 0) {
+        return S2N_SUCCESS;
+    }
+    int base64_data_size = base64_groups * 4;
+    int binary_output_size = base64_groups * 3;
+
+    const uint32_t base64_data_offset = stuffer->read_cursor;
+    POSIX_GUARD(s2n_stuffer_skip_read(stuffer, base64_data_size));
+    const uint8_t *start_of_base64_data = stuffer->blob.data + base64_data_offset;
+
+    const uint32_t binary_output_offset = out->write_cursor;
+    POSIX_GUARD(s2n_stuffer_skip_write(out, binary_output_size));
+    uint8_t *start_of_binary_output = out->blob.data + binary_output_offset;
+
+    /* https://docs.openssl.org/master/man3/EVP_EncodeInit/
+     * > This function will return the length of the data decoded or -1 on error. */
+    int res = EVP_DecodeBlock(start_of_binary_output, start_of_base64_data, base64_data_size);
+    POSIX_ENSURE(res == binary_output_size, S2N_ERR_INVALID_BASE64);
+
+    /* https://docs.openssl.org/1.1.1/man3/EVP_EncodeInit/
+     * > The output will be padded with 0 bits if necessary to ensure that the 
+     * > output is always 3 bytes for every 4 input bytes. 
+     * FFFF -> 0x14 0x51 0x45
+     * FFF= -> 0x14 0x51 0x00
+     * FF== -> 0x14 0x00 0x00
+     * F=== -> INVALID
+     */
+    /* manually unrolled loop to prevent CBMC errors */
+    if (stuffer->blob.data[stuffer->read_cursor - 1] == '=') {
+        out->write_cursor -= 1;
+    }
+    if (stuffer->blob.data[stuffer->read_cursor - 2] == '=') {
+        out->write_cursor -= 1;
+    }
```

Comment:
Nit: I know your "base64_groups == 0" check above means that read_cursor >= 4 and the CBMC proof should catch any mistakes like this, but it might still be worth an explicit ENSURE(stuffer->read_cursor > 2) check right before you try to read these indexes, just to make the precondition very clear.